### PR TITLE
[Runtime] Add a disabled workaround for protocol conformance checking to check conformances in reverse order.

### PR DESCRIPTION
**Pull request by mikeash**
rdar://problem/72049977

### CODE REVIEW

**Comment by mikeash** [comment]
@swift-ci please test

**Comment by mikeash** [comment]
@swift-ci please test windows platform

File: stdlib/public/runtime/Bincompat.cpp [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+//===----------------------------------------------------------------------===//
+//
+// Checks for enabling binary compatibility workarounds.
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/Runtime/Bincompat.h"
+
+bool swift::runtime::bincompat::
+    workaroundProtocolConformanceReverseIteration() {
```
Comment (compnerd) [inline stdlib/public/runtime/Bincompat.cpp:20]: Might be nice to do the following for wrapping (or just clang-format it):

```c++
namespace swift {
namespace runtime {
namespace bincompat {
bool workaroundProtocolConformanceReverseIteraction() {
  return false;
}
}
}
}
```

File: stdlib/public/runtime/ProtocolConformance.cpp [hunk 2]
```diff
@@ -238,8 +239,11 @@ namespace {
 struct ConformanceState {
   ConcurrentReadableHashMap<ConformanceCacheEntry> Cache;
   ConcurrentReadableArray<ConformanceSection> SectionsToScan;
-  
+  bool scanSectionsBackwards;
+
   ConformanceState() {
+    scanSectionsBackwards = swift::runtime::bincompat::
```
Comment (compnerd) [inline stdlib/public/runtime/ProtocolConformance.cpp:245]: Might be able to get away with `bincompat::workaroundProtocolConformanceReverseIteration()`.

File: stdlib/public/runtime/ProtocolConformance.cpp [hunk 3]
```diff
@@ -485,6 +488,18 @@ swift_conformsToProtocolImpl(const Metadata *const type,
         C.cacheResult(matchingType, protocol, witness, /*always cache*/ 0);
       }
     }
+  };
+
+  auto snapshot = C.SectionsToScan.snapshot();
+  if (C.scanSectionsBackwards) {
```
Comment (compnerd) [inline stdlib/public/runtime/ProtocolConformance.cpp:494]: Might be nice to do something like:

```
if (C.scanSectionsBackwards)
  std::reverse(std::begin(snapshot), std::end(snapshot))
for (auto &section : snapshot)
  processSection(section);
```

File: stdlib/public/runtime/Bincompat.cpp (hunk 1 above)
Comment (mikeash) [reply stdlib/public/runtime/Bincompat.cpp]: Clang-format gave me the version I had, but your version is still nicer. Updated.

File: stdlib/public/runtime/ProtocolConformance.cpp (hunk 2 above)
Comment (mikeash) [reply stdlib/public/runtime/ProtocolConformance.cpp]: Almost. `runtime::bincompat::...` works. Updated.

File: stdlib/public/runtime/ProtocolConformance.cpp (hunk 3 above)
Comment (mikeash) [reply stdlib/public/runtime/ProtocolConformance.cpp:494]: As discussed offline, we can't mutate in place, but we settled on:
```
  auto snapshot = C.SectionsToScan.snapshot();
  if (C.scanSectionsBackwards) {
    for (auto &section : llvm::reverse(snapshot))
      processSection(section);
  } else {
    for (auto &section : snapshot)
      processSection(section);
  }
```

**Comment by mikeash** [comment]
@swift-ci please test

**Review by compnerd — APPROVED** [review]
Thanks!

**Comment by mikeash** [comment]
@swift-ci please test

**Pull request merged**